prov/gni/src/gnix_wait.c

```diff
@@ -30,9 +30,159 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-
+#include <stdlib.h>
+#include <signal.h>
 #include "gnix.h"
 #include "gnix_wait.h"
+#include "gnix_nic.h"
+
+/*
+ * Gnix wait progress thread declarations for making sure nic progress
+ * occurs when inside a gnix_wait call
+ */
+
+static pthread_t	gnix_wait_thread;
+static pthread_mutex_t	gnix_wait_mutex = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t	gnix_wait_cond;
+static volatile int	gnix_wait_thread_ready;
```

Comment:
Replace the and following volatiles with atomics and appropriate functions on atomics.  The behavior of volatiles is not defined in the memory consistency model.